### PR TITLE
Add a working Formatter for Rholang-idea

### PR DESCRIPTION
**Pull request by AbnerZheng**
This PR help add a working formatter for the plugin. 
There is still some things to improve. But I think it is time to gather feedback from users. 

The function preview of this PR:
image

### CODE REVIEW

File: src/coop/rchain/lang/Rho.bnf [hunk 1]
```diff
@@ -106,23 +106,28 @@ private stringLit ::= STRING_LIT | RAW_STRING_LIT | CHAR_LIT
...
+        | ConstrName ParenthesizedProc
+        | ID_NAME ParenthesizedProc
         | Contract_
         | "@" ID_NAME
-Proc_ ::= (FnDef | Proc1 | Proc2 | Proc3 | Proc4 | ConstrName) [TypePattern] | comment
+
+ParenthesizedProc::= "(" [Proc] ")"
+MatchWith ::= MATCH Proc WITH [PMBranch]
+
+ForParamsList ::= "(" [Bind] ")"
```
Comment (tgrospic) [inline src/coop/rchain/lang/Rho.bnf]: You already replaced `"{" [Proc] "}"` with `ProcBlock`. It looks like a similar situation now but with parenthesis and very possible for other cases in the future. I'm thinking that we can have a naming rule like:
```
ProcBlock  ::= "{" [Proc] "}"
ProcParen ::= "(" [Proc] ")"
BindParen ::= "(" [Bind] ")"
```
And also `CBranchBlock ::= "{" [CBranch] "}"` to support formatting of select-case block.

File: src/coop/rchain/lang/Rho.bnf [hunk 2]
```diff
@@ -173,7 +178,7 @@ QVal ::= TRUE
 
 // Quantity
 Quantity_ ::= ("-"|"+") Quantity_
-  | "(" [Proc] ")"
+  | ParenthesizedProc
   | "[" [Proc] "]"
   | "{" [Proc] "}"
```
Comment (tgrospic) [inline src/coop/rchain/lang/Rho.bnf]: We missed this the last time. It is also `ProcBlock`.

File: testData/ide/formatter/tokenFormatted.rho [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+                        new x, y in {
+                            x!(balances.getOrElse(message_sender, 0)) |
+                            y!(balances.getOrElse(to, 0)) |
+                            for ( balance_of_sender <- x; balance_of_receiver <- y; balacen_of_test<-z) {
+                                // Preconditions to check before the transfer goes through
+                                match [
+                                  balance_of_sender >= amount,
+                                  amount > 0,
+                                  balance_of_receiver + amount > balance_of_receiver
+                                ] with
```
Comment (tgrospic) [inline testData/ide/formatter/tokenFormatted.rho]: Auto format will always put `] with` in a new line. It doesn't looks nice if `match [...] with` is in one line. Is this easily fixable?

File: src/coop/rchain/lang/formatter/RholangIndentProcessor.kt [hunk 4]
```diff
@@ -0,0 +1,52 @@
+package coop.rchain.lang.formatter
+
+import a.d.P
```
Comment (tgrospic) [inline src/coop/rchain/lang/formatter/RholangIndentProcessor.kt]: This import is not recognized.

File: .idea/codeStyleSettings.xml [hunk 5]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (tgrospic) [inline .idea/codeStyleSettings.xml:1]: Please pull the latest master, I removed `.idea` folder completely after your last PR with folding. The project already contains `.editorconfig` to specify formatting settings so maybe it's better not to have both.

File: testData/ide/formatter/tokenFormatted.rho (hunk 3 above)
Comment (AbnerZheng) [reply testData/ide/formatter/tokenFormatted.rho]: Yes, it can be fixed.

File: src/coop/rchain/lang/formatter/RholangBlock.kt [hunk 6]
```diff
@@ -0,0 +1,61 @@
+package coop.rchain.lang.formatter
```
Comment (tgrospic) [inline src/coop/rchain/lang/formatter/RholangBlock.kt:1]: Please reformat code files to have indentation specified in `.editorconfig` with 2 spaces. Later modifications will be much cleaner.

Although on [KotlinLang](https://kotlinlang.org/docs/reference/coding-conventions.html#formatting) they say it's usually 4 spaces. What do you think, should we switch to 4 spaces? It's the similar question for Rholang.

File: testData/ide/formatter/simpleOneFormatted.rho [hunk 7]
```diff
@@ -0,0 +1,7 @@
+new time_contract in {
```
Comment (tgrospic) [inline testData/ide/formatter/simpleOneFormatted.rho:1]: The test is failing for this file, actual file has 4 spaces.

**Review by tgrospic — CHANGES_REQUESTED** [review]
Nice work, please see my comments before I merge your changes.

After merge I will update bnf syntax with the latest from RChain repo.

**Comment by AbnerZheng** [comment]
Sorry for replying later. I was in Spring Festival and busy with other issue. Now I will have some time to do with this issue.

File: src/coop/rchain/lang/Rho.bnf (hunk 1 above)
Comment (AbnerZheng) [reply src/coop/rchain/lang/Rho.bnf]: Aha, actually I have these BindParen and ProcParen with different name, as ForParamsList ::= "(" [Bind] ")" and  ParenthesizedProc::= "(" [Proc] ")".  Your suggestion is more consistent.

File: src/coop/rchain/lang/Rho.bnf (hunk 2 above)
Comment (AbnerZheng) [reply src/coop/rchain/lang/Rho.bnf]: Yeah, you're right.

File: src/coop/rchain/lang/formatter/RholangBlock.kt (hunk 6 above)
Comment (AbnerZheng) [reply src/coop/rchain/lang/formatter/RholangBlock.kt:1]: It is ok for 2 spaces when in KotlinLang. 
And for Rholang, I will create a new PR for add setting support for these preferred style.

File: testData/ide/formatter/simpleOneFormatted.rho (hunk 7 above)
Comment (AbnerZheng) [reply testData/ide/formatter/simpleOneFormatted.rho:1]: I have fixed the text file. 
Two or four spaces should be set by users. Wait for my further PR. : )

**Comment by AbnerZheng** [comment]
@tgrospic Please have a look.

**Comment by tgrospic** [comment]
Sorry for my late response.

Formatting is working as expected, except when the code is not highlighted (by some syntax error) formatting removes spaces on every line start. Can you check how to disable formatting if the code is not successfully parsed?

I also noticed that your PR now includes my two commits. Please sync your repo and rebase your PR with the current master on my repo. I added these two commits after your last PR with folding feature.

**Comment by AbnerZheng** [comment]
@tgrospic 
hi, the right way to do with syntax error is using  error recovery provided by BNF Grammars, see https://github.com/JetBrains/Grammar-Kit#attributes-for-error-recovery-and-reporting .  When doing this way, the parser will always generate a working AST even there are some error. I am not familiar with this, but I will figure out how to tune the parser and help to do these things. 
I will rebase the PR today, and I think we can merge the PR firstly and publish this feature after we have update the grammar to 0.2 and doing something about error recovery.

**Comment by AbnerZheng** [comment]
@tgrospic 
I have rebased the master. Please have a look.

**Comment by tgrospic** [comment]
@AbnerZheng to rebase your repo, you will have to rewrite the repo history. This will pull changes from my repo and rewrite yours on top. You will have one conflict for the deleted file.

`git pull --rebase -s recursive -X ours git@github.com:tgrospic/rholang-idea.git`

Since history has changed, you must use the `--force` parameter when pushing your repo.

**Comment by AbnerZheng** [comment]
Hi, is it ok now? I have done what you suggested.

**Comment by tgrospic** [comment]
Everything looks good. @AbnerZheng thanks for the formatting improvements. :smiley: :tada: 

After merging your changes, I will update the grammar and publish a new version.

**Review by tgrospic — APPROVED** [review]

**Pull request merged**